polynote-kernel/src/main/scala/polynote/kernel/dependency/CoursierFetcher.scala

```diff
@@ -0,0 +1,206 @@
+    case repo @ ivy(base, _, _, changing) =>
+      val baseUri = base.stripSuffix("/") + "/"
+      val artifactPattern = s"$baseUri${repo.artifactPattern}"
+      val metadataPattern = s"$baseUri${repo.metadataPattern}"
+      Validated.fromEither(IvyRepository.parse(artifactPattern, Some(metadataPattern), changing = changing)).toValidatedNel
+    case maven(base, changing) =>
+      val repo = MavenRepository(base, changing = changing)
+      Validated.validNel(repo)
+
+  }.sequence[ValidatedNel[String, ?], Repository].leftMap {
+    errs => new RuntimeException(s"Errors parsing repositories:\n- ${errs.toList.mkString("\n- ")}")
+  }.toEither
+
+
+  private def resolveCoursierDependencies(
+    resolution: Resolution,
+    taskInfo: TaskInfo,
+    statusUpdates: Publish[IO, KernelStatusUpdate],
+    maxIterations: Int = 100
+  ) = {
+    // check whether we care about a missing resolution
+    def shouldErrorIfMissing(mv: (Module, String)): Boolean = {
+      resolution.rootDependencies.exists(dep => dep.module == mv._1 && dep.version == mv._2)
+    }
+
+    def publish(name: String, progress: Double) = statusUpdates.publish1(UpdatedTasks(List(TaskInfo(name, name.split('/').last, name, TaskStatus.Running, (progress * 255).toByte))))
+    def complete(name: String) = statusUpdates.publish1(UpdatedTasks(List(TaskInfo(name, "", "", TaskStatus.Complete, 255.toByte))))
+    val cache = this.cache.withLogger {
+      new CacheLogger {
+        private var totalArtifacts: Int = 0
+        private val knownLength = new ConcurrentHashMap[String, Long]()
+        override def init(sizeHint: Option[Int]): Unit = totalArtifacts = sizeHint.getOrElse(resolution.minDependencies.size)
+
+        override def downloadingArtifact(url: String): Unit = publish(url, 0.0).unsafeRunAsyncAndForget()
+        override def downloadLength(url: String, totalLength: Long, alreadyDownloaded: Long, watching: Boolean): Unit = {
+          knownLength.put(url, totalLength)
+          if (watching && alreadyDownloaded < totalLength) {
+            publish(url, alreadyDownloaded.toDouble / totalLength).unsafeRunAsyncAndForget()
+          }
+        }
+
+        override def downloadProgress(url: String, downloaded: Long): Unit = {
+          if (knownLength.containsKey(url)) {
+            val len = knownLength.get(url)
+            if (downloaded < len) {
+              publish(url, downloaded.toDouble / len).unsafeRunAsyncAndForget()
+            } else {
+              complete(url).unsafeRunAsyncAndForget()
+            }
+          }
+        }
+
+        override def downloadedArtifact(url: String, success: Boolean): Unit = {
+          complete(url).unsafeRunAsyncAndForget()
+        }
+      }
+    }
+
+    Artifacts(cache).withResolution(resolution).withMainArtifacts(true).io.map {
+      artifacts => artifacts.toList.map {
+        case (artifact, file) => artifact.url -> IO.pure(file)
```

Comment:
So this is a bit odd... we have to return `List[(String, IO[File])]` but their API wraps the whole thing in one `IO`. So by the time we return the "tasks", the download is already done. It doesn't make a big difference, just pointing out why this might look a little strange.